jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -411,30 +422,36 @@ public void onFillable()
                     if (_handling.compareAndSet(true, false))
                     {
                         if (LOG.isDebugEnabled())
-                            LOG.debug("request !complete {} {}", request, this);
+                            LOG.debug("request !complete {} {} {}", request, this, _requestBuffer);
+                        releaseRequestBufferIfEmpty();
```

Comment:
This is in a race with the application, so we cannot release at all here.
A thread spawned by the application can read request content, possibly leading to unexpected NPEs.